### PR TITLE
change address method to return list of addresses

### PR DESCRIPTION
**Pull request by aykut-bozkurt**
We change address method's return value for distributed objects. It returns list of object adresses instead of single address because some of the statements accepts more than 1 object.

### CODE REVIEW

File: src/backend/distributed/commands/function.c [hunk 1]
```diff
@@ -1373,27 +1373,35 @@ PostprocessCreateFunctionStmt(Node *node, const char *queryString)
 		return NIL;
 	}
 
-	ObjectAddress functionAddress = GetObjectAddressFromParseTree((Node *) stmt, false);
+	List *commands = list_make1(DISABLE_DDL_PROPAGATION);
+
+	List *functionAddresses = GetObjectAddressFromParseTree((Node *) stmt, false);
+	Assert(list_length(functionAddresses) == 1);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/function.c:1379]: I think where we add `Assert(list_length(functionAddresses) == 1);` we should not change the code to iterate over the list

File: src/backend/distributed/commands/common.c [hunk 2]
```diff
@@ -61,22 +61,24 @@ PostprocessCreateDistributedObjectFromCatalogStmt(Node *stmt, const char *queryS
 		return NIL;
 	}
 
-	ObjectAddress address = GetObjectAddressFromParseTree(stmt, false);
+	List *addresses = GetObjectAddressFromParseTree(stmt, false);
+	Assert(list_length(addresses) == 1);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/common.c:65]: I think for the places where we have `Assert` we need 1 line comment:

```
/*  the code-path only supports a single object */
Assert(list_length(addresses) == 1);
```

or such

File: src/backend/distributed/commands/common.c [hunk 3]
```diff
@@ -111,8 +113,10 @@ PreprocessAlterDistributedObjectStmt(Node *stmt, const char *queryString,
 	const DistributeObjectOps *ops = GetDistributeObjectOps(stmt);
 	Assert(ops != NULL);
 
-	ObjectAddress address = GetObjectAddressFromParseTree(stmt, false);
-	if (!ShouldPropagateObject(&address))
+	List *addresses = GetObjectAddressFromParseTree(stmt, false);
+	Assert(list_length(addresses) == 1);
+
+	if (!ShouldPropagateObjects(addresses))
```
Comment (onderkalaci) [inline src/backend/distributed/commands/common.c:119]: I think `ShouldPropagateAnyObject` is easier to understand what the function does

File: src/backend/distributed/commands/dependencies.c [hunk 4]
```diff
@@ -681,6 +707,22 @@ ShouldPropagateObject(const ObjectAddress *address)
 }
 
 
+bool
+ShouldPropagateObjects(List *addresses)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/dependencies.c:711]: As noted, lets update the function name to `ShouldPropagateAnyObject`


Also, all new functions need function comment

File: src/include/distributed/metadata_utility.h [hunk 5]
```diff
@@ -259,14 +259,17 @@ extern void CreateTruncateTrigger(Oid relationId);
 extern TableConversionReturn * UndistributeTable(TableConversionParameters *params);
 
 extern void EnsureDependenciesExistOnAllNodes(const ObjectAddress *target);
```
Comment (onderkalaci) [inline src/include/distributed/metadata_utility.h:261]: EnsureDependenciesExistOnAllNodes and all other internal functions should become `static` functions defined locally in C files. We should only expose the APIs which takes the list

File: src/backend/distributed/commands/function.c [hunk 6]
```diff
@@ -1373,27 +1393,29 @@ PostprocessCreateFunctionStmt(Node *node, const char *queryString)
...
 		return NIL;
 	}
 
-	EnsureDependenciesExistOnAllNodes(&functionAddress);
+	EnsureAllObjectDependenciesExistOnAllNodes(functionAddresses);
 
 	List *commands = list_make1(DISABLE_DDL_PROPAGATION);
-	commands = list_concat(commands, CreateFunctionDDLCommandsIdempotent(
-							   &functionAddress));
+	commands = list_concat(commands, CreateAllFunctionDDLCommandsIdempotent(
```
Comment (onderkalaci) [inline src/backend/distributed/commands/function.c:1417]: Ok, I keep getting here. But, I'd rather drop `CreateAllFunctionDDLCommandsIdempotent` function, and do

```
/* we already asserted above that there is a single */
ObjectAddress *functionAddress = (ObjectAddress *) linitial(functionAddresses);

... and use CreateFunctionDDLCommandsIdempotent ... 
```


I suggest this because `CreateAllFunctionDDLCommandsIdempotent` is really not a generic function. We are sure that there is no caller that can call this with multiple objects.

File: src/backend/distributed/commands/function.c [hunk 7]
```diff
@@ -1494,8 +1516,10 @@ PreprocessAlterFunctionStmt(Node *node, const char *queryString,
 	AlterFunctionStmt *stmt = castNode(AlterFunctionStmt, node);
 	AssertObjectTypeIsFunctional(stmt->objtype);
 
-	ObjectAddress address = GetObjectAddressFromParseTree((Node *) stmt, false);
-	if (!ShouldPropagateAlterFunction(&address))
+	List *addresses = GetObjectAddressFromParseTree((Node *) stmt, false);
+	Assert(list_length(addresses) == 1);
+
+	if (!ShouldPropagateAlterFunction(addresses))
```
Comment (onderkalaci) [inline src/backend/distributed/commands/function.c:1522]: similar to `CreateAllFunctionDDLCommandsIdempotent `, I'd rather drop very specific functions such as `ShouldPropagateAlterFunction` which ever only accepts a single address.

Postgres syntax does not allow altering multiple functions in a single statement anyway.

File: src/backend/distributed/metadata/dependency.c [hunk 8]
```diff
@@ -900,6 +916,24 @@ DeferErrorIfHasUnsupportedDependency(const ObjectAddress *objectAddress)
 }
 
 
+DeferredErrorMessage *
+DeferErrorIfAnyObjectHasUnsupportedDependency(const List *objectAddresses)
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/dependency.c:920]: needs comment, try to make `DeferErrorIfHasUnsupportedDependency` static

File: src/backend/distributed/metadata/distobject.c [hunk 9]
```diff
@@ -393,8 +393,8 @@ UnmarkObjectDistributed(const ObjectAddress *address)
 bool
 IsObjectDistributed(const ObjectAddress *address)
 {
-	ScanKeyData key[3];
 	bool result = false;
+	ScanKeyData key[3];
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/distobject.c:397]: can you undo the changes in this file? they seem irrelevant?

**Review by onderkalaci — COMMENTED** [review]
Sorry for the multiple iterations on this commit. Now, looking at the patch, I think my general inclination is (a) For generic APIs like  `ErrorOrWarnIfAnyObjectHasUnsupportedDependency,  EnsureAllObjectDependenciesExistOnAllNodes, IsAnyObjectAddressOwnedByExtension, ShouldPropagateObjects, IsAnyObjectDistributed`, lets get a `List` (b) for very specific apis, like `CreateAllFunctionDDLCommandsIdempotent `, we do not iterate

We are very close to merge, majority of my comments are style

File: src/backend/distributed/commands/dependencies.c [hunk 10]
```diff
@@ -681,6 +719,26 @@ ShouldPropagateObject(const ObjectAddress *address)
...
+ * ShouldPropagateAnyObject determines if we should be propagating DDLs based
+ * on their object addresses.
+ */
+bool
+ShouldPropagateAnyObject(List *addresses)
+{
+	ObjectAddress *address = NULL;
+	foreach_ptr(address, addresses)
+	{
+		if (!ShouldPropagateObject(address))
```
Comment (onderkalaci) [inline src/backend/distributed/commands/dependencies.c:732]: should we return true `if(ShouldPropagateObject)` ? `ShouldPropagateAnyObject` implies that, similar to IsAnyObjectDistributed

File: src/backend/distributed/commands/function.c [hunk 11]
```diff
@@ -1562,7 +1581,7 @@ PreprocessAlterFunctionDependsStmt(Node *node, const char *queryString,
 	 */
 
 	const char *functionName =
-		getObjectIdentity_compat(&address, /* missingOk: */ false);
+		getObjectIdentity_compat(linitial(addresses), /* missingOk: */ false);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/function.c:1584]: minor: maybe for ease of read, define `ObjectAddress *address =linital..`

File: src/backend/distributed/commands/sequence.c [hunk 12]
```diff
@@ -747,7 +772,7 @@ PostprocessGrantOnSequenceStmt(Node *node, const char *queryString)
 		ObjectAddress sequenceAddress = { 0 };
 		Oid sequenceOid = RangeVarGetRelid(sequence, NoLock, false);
 		ObjectAddressSet(sequenceAddress, RelationRelationId, sequenceOid);
-		EnsureDependenciesExistOnAllNodes(&sequenceAddress);
+		EnsureAllObjectDependenciesExistOnAllNodes(list_make1(&sequenceAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/commands/sequence.c:775]: I don't think we should ever pass a reference of a local variable to a list. The lifetime of `sequenceAddress` is scoped to the end of `PostprocessGrantOnSequenceStmt`.

Here, we are not in danger, because the list is used within `PostprocessGrantOnSequenceStmt`. However, as a better practice, please allocate the object addresses via palloc, and set its values, such as in `GetDistributedObjectAddressList`.

Can you go over your patch to see if we have any other such cases? I can see at least `PostprocessAlterStatisticsOwnerStmt `

File: src/backend/distributed/worker/worker_drop_protocol.c [hunk 13]
```diff
@@ -177,7 +177,7 @@ WorkerDropDistributedTable(Oid relationId)
 	 * until the user runs DROP EXTENSION. Therefore, we skip dropping the
 	 * table.
 	 */
-	if (!IsObjectAddressOwnedByExtension(&distributedTableObject, NULL))
+	if (!IsAnyObjectAddressOwnedByExtension(list_make1(&distributedTableObject), NULL))
```
Comment (onderkalaci) [inline src/backend/distributed/worker/worker_drop_protocol.c:180]: would be good to avoid passing reference to a local variable again, maybe convert distributedTableObject to a pointer

**Review by onderkalaci — COMMENTED** [review]
we probably need the last iteration before the merge, thanks

File: src/backend/distributed/commands/alter_table.c [hunk 14]
```diff
@@ -1318,7 +1318,9 @@ DoesCascadeDropUnsupportedObject(Oid classId, Oid objectId, HTAB *nodeMap)
 	ObjectAddress objectAddress = { 0 };
 	ObjectAddressSet(objectAddress, classId, objectId);
 
-	if (IsObjectAddressOwnedByExtension(&objectAddress, NULL))
+	ObjectAddress *copyObjectAddress = palloc0(sizeof(ObjectAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/commands/alter_table.c:1321]: Why do we need `copyObjectAddress`? Can't we do;
```C
ObjectAddress *objectAddress = palloc0(sizeof(ObjectAddress));
ObjectAddressSet(*objectAddress, classId, objectId)
```

File: src/backend/distributed/commands/extension.c [hunk 15]
```diff
@@ -322,7 +325,9 @@ FilterDistributedExtensions(List *extensionObjectList)
 		ObjectAddress address = { 0 };
 		ObjectAddressSet(address, ExtensionRelationId, extensionOid);
 
-		if (!IsObjectDistributed(&address))
+		ObjectAddress *copyAddress = palloc0(sizeof(ObjectAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/commands/extension.c:328]: can we avoid `copyAddress`?

File: src/backend/distributed/commands/foreign_data_wrapper.c [hunk 16]
```diff
@@ -90,9 +91,14 @@ NameListHasFDWOwnedByDistributedExtension(List *FDWNames)
 		ObjectAddress extensionAddress = { 0 };
 		ObjectAddress FDWAddress = GetObjectAddressByFDWName(strVal(FDWValue), false);
 
-		if (IsObjectAddressOwnedByExtension(&FDWAddress, &extensionAddress))
+		ObjectAddress *copyFDWAddress = palloc0(sizeof(ObjectAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/commands/foreign_data_wrapper.c:94]: can we avoid copyFDWAddress and copyExtensionAddress?

File: src/backend/distributed/commands/function.c [hunk 17]
```diff
@@ -224,7 +224,10 @@ create_distributed_function(PG_FUNCTION_ARGS)
 	 * pg_dist_object, and not propagate the CREATE FUNCTION. Function
 	 * will be created by the virtue of the extension creation.
 	 */
-	if (IsObjectAddressOwnedByExtension(&functionAddress, &extensionAddress))
+	ObjectAddress *copyFunctionAddress = palloc0(sizeof(ObjectAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/commands/function.c:227]: can we avoid `copyFunctionAddress` ? It is simpler to adjust `EnsureExtensionFunctionCanBeDistributed` to get a pointer

File: src/backend/distributed/commands/grant.c [hunk 18]
```diff
@@ -240,7 +240,10 @@ CollectGrantTableIdList(GrantStmt *grantStmt)
 			/* check for distributed sequences included in GRANT ON TABLE statement */
 			ObjectAddress sequenceAddress = { 0 };
```
Comment (onderkalaci) [inline src/backend/distributed/commands/grant.c:241]: can you avoid `copySequenceAddress`?

File: src/backend/distributed/commands/index.c [hunk 19]
```diff
@@ -763,7 +763,10 @@ PostprocessIndexStmt(Node *node, const char *queryString)
 	/* ensure dependencies of index exist on all nodes */
 	ObjectAddress address = { 0 };
 	ObjectAddressSet(address, RelationRelationId, indexRelationId);
-	EnsureDependenciesExistOnAllNodes(&address);
+
+	ObjectAddress *copyAddress = palloc0(sizeof(ObjectAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/commands/index.c:767]: can we avoid copyAddress

File: src/backend/distributed/commands/statistics.c [hunk 20]
```diff
@@ -452,7 +458,9 @@ PostprocessAlterStatisticsOwnerStmt(Node *node, const char *queryString)
 	ObjectAddress statisticsAddress = { 0 };
 	ObjectAddressSet(statisticsAddress, StatisticExtRelationId, statsOid);
 
-	EnsureDependenciesExistOnAllNodes(&statisticsAddress);
+	ObjectAddress *copyStatisticsAddress = palloc0(sizeof(ObjectAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/commands/statistics.c:461]: do we really need copyStatisticsAddress?

File: src/backend/distributed/commands/text_search.c [hunk 21]
```diff
@@ -825,16 +825,22 @@ TextSearchDictCommentObjectAddress(Node *node, bool missing_ok)
...
 AlterTextSearchConfigurationOwnerObjectAddress(Node *node, bool missing_ok)
 {
 	AlterOwnerStmt *stmt = castNode(AlterOwnerStmt, node);
 	Relation relation = NULL;
 
 	Assert(stmt->objectType == OBJECT_TSCONFIGURATION);
 
-	return get_object_address(stmt->objectType, stmt->object, &relation, AccessShareLock,
-							  missing_ok);
+	ObjectAddress objectAddress = get_object_address(stmt->objectType, stmt->object,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/text_search.c:836]: for example, here it is needed to have `objectAddressCopy`  because we cannot pass `&objectAddress`

File: src/backend/distributed/metadata/metadata_sync.c [hunk 22]
```diff
@@ -359,7 +359,9 @@ CreateDependingViewsOnWorkers(Oid relationId)
 		ObjectAddress viewAddress = { 0 };
 		ObjectAddressSet(viewAddress, RelationRelationId, viewOid);
 
-		EnsureDependenciesExistOnAllNodes(&viewAddress);
+		ObjectAddress *copyViewAddress = palloc0(sizeof(ObjectAddress));
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_sync.c:362]: do we really need `copyViewAddress`? Can't we make viewAddress as a pointer?

**Review by onderkalaci — COMMENTED** [review]
pretty minor comments

File: src/backend/distributed/commands/foreign_data_wrapper.c (hunk 16 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/foreign_data_wrapper.c:94]: Only copyExtensionAddress

**Review by onderkalaci — APPROVED** [review]

File: src/backend/distributed/commands/type.c [hunk 23]
```diff
@@ -117,8 +117,12 @@ PreprocessRenameTypeAttributeStmt(Node *node, const char *queryString,
 	Assert(stmt->renameType == OBJECT_ATTRIBUTE);
 	Assert(stmt->relationType == OBJECT_TYPE);
 
-	ObjectAddress typeAddress = GetObjectAddressFromParseTree((Node *) stmt, false);
-	if (!ShouldPropagateObject(&typeAddress))
+	List *typeAddresses = GetObjectAddressListFromParseTree((Node *) stmt, false);
+
+	/*  the code-path only supports a single object */
+	Assert(list_length(objectAddresses) == 1);
```
Comment (naisila) [inline src/backend/distributed/commands/type.c:123]: should be `typeAddresses` here

File: src/backend/distributed/commands/table.c [hunk 24]
```diff
@@ -649,13 +649,19 @@ PostprocessAlterTableSchemaStmt(Node *node, const char *queryString)
 	/*
 	 * We will let Postgres deal with missing_ok
 	 */
-	ObjectAddress tableAddress = GetObjectAddressFromParseTree((Node *) stmt, true);
+	List *tableAddresses = GetObjectAddressListFromParseTree((Node *) stmt, true);
+
+	/*  the code-path only supports a single object */
+	Assert(list_length(tableAddress) == 1);
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:655]: ops, good point

File: src/backend/distributed/worker/worker_create_or_replace.c [hunk 25]
```diff
@@ -181,8 +181,13 @@ WorkerCreateOrReplaceObject(List *sqlStatements)
 	 * same subject.
 	 */
 	Node *parseTree = ParseTreeNode(linitial(sqlStatements));
-	ObjectAddress address = GetObjectAddressFromParseTree(parseTree, true);
-	if (ObjectExists(&address))
+	List *addresses = GetObjectAddressListFromParseTree(parseTree, true);
+	Assert(list_length(viewAddresses) == 1);
```
Comment (naisila) [inline src/backend/distributed/worker/worker_create_or_replace.c:185]: should be `addresses`
